Update installation pages (#1273)
PelicanPlatform/pelican: docs/pages/install/docker.mdx

# Run Pelican Server with Docker Image

> This document explains how to run a Pelican server using Pelican docker image. If you want to use Pelican client functionalities, such as download or upload an object, refer to [previous sections](../install.mdx) to download and install a Pelican binary instead.

Reviewer: Currently this comment suggests that you can't use the Docker image for client functionality; I think this is incorrect.

```suggestion
This document explains how to run a Pelican server using Pelican docker image. For Pelican client functionalities, such as download or upload an object, you can use the docker image as well, or you can install the Pelican binary instead as described [elsewhere](../install.mdx).
```

Author: They can, but our Docker images aren't set up in a way to facilitate client usage, and we don't recommend this behavior.

Reviewer: Ah, got it. 
Totally did not override the entrypoint of the container, add pelican to the PATH, and use that as a client to download a test file. That would be silly..

* Pelican director server: `hub.opensciencegrid.org/pelican_platform/director:latest`
* Pelican registry server: `hub.opensciencegrid.org/pelican_platform/registry:latest`

> > **Note**: the `latest` tag will pull the Pelican server image with the latest released version. You may pull an image with explicit Pelican version by passing the version as the tag (e.g. `v7.6.4`) instead. For a list of available tags, refer to [Pelican Harbor repository](https://hub.opensciencegrid.org/harbor/projects/883/repositories/origin/artifacts-tab?publicAndNotLogged=yes).

Reviewer 2: This Harbor URL brings users to a page with a big scary red banner that says "Unauthorized". Presumably we have an image for every tagged/released version, so maybe we can point them to the github Releases page instead?

Reviewer: I'm also not a fan of the scary red banner. That probably requires something to change on the Harbor side. 

Maybe link to a document (in Github?) that tables the versions and the Harbor tags / docker addresses? That way people can find the docker addresses without encountering the scary red banner.

Author: I'd suggest change the GitHub release note to include links to various Docker images, or have a page on our docs/Pelican website listing available containers. @CannonLock What's your thought on this?

Where most of the command overlaps with the one to run the Pelican origin server, with the following differences:
* `-p 8444:8444 -p 8442:8442` publishes port `8444` and `8442` from the container to the same port on the host machine. Note that the admin website of the Pelican server runs on port `8444` by default, and the objects transfer endpoint of the Pelican cache server runs on port `8442` by default.

> > **Note**: Pelican Docker image currently does not support binding a directory on your host machine as the directory for Pelican cache to store cache files, i.e. using `-v /host/machine:/run/pelican/cache/location`. We will have an update once we officialy support this flow.

Reviewer 3: ```suggestion
> **Note**: the Pelican Docker image currently does not support binding a directory on your host machine as the directory for a Pelican cache to store cached objects, i.e. using `-v /host/machine:/run/pelican/cache/location`. We will have an update once we officially support this flow.
```

And instead of "We will have an update once we officially support this flow." Maybe say something like this:
"This feature is a work in progress and will be available in a future Pelican version"

This sounds good to me but maybe others have another suggestion as well.

Reviewer 2: Following up on Joe's `e.g` versus `i.e.` comment, they mean two different things and aren't interchangeable so we can't pick one over the other. The way I remember which to use is that "i.e." is Latin for "id est" and means "that is" whereas "e.g." is Latin for "exempli gratia" and means "for example". The former is to clarify something when there's only one way it can be clarified, and the latter is for any time you can construct an arbitrary example. In this case I think "e.g." is correct, but things look correct in most of the other usage.

Author: Thanks Justin for the explanation!

docker run -it -p 8444:8444 --name=pelican-registry hub.opensciencegrid.org/pelican_platform/registry:latest serve -f <federation>
```

> To successfully run a Pelican registry server, additional configuration is required. Follow [Serve a Pelican Registry](../serving_a_federation.mdx#serve-a-registry) for instructions. For how to pass configurations to Docker container, refer to [the next section](#configure-pelican-server-in-a-container).

Reviewer: Should this statement go before you show the example command for running the director container? 
My concern is that people will see and try to run the command before reading the note that it won't work as it is.

4. You should note that your running Pelican origin server is logging debug level messages.

## Additional Configurations for Running Pelican in a Container

### Server hostname and port

> Pelican server has a built-in web server to handle various API requests. By default, it uses server hostname and a port number as the server address (e.g. `https://127.0.0.1:8444`). In a container environment, the hostname is an alphanumeric value, e.g. `6ee28c5df997`, and the corresponding Pelican server address is `https://6ee28c5df997:8444`. This value is for container internal communications and is not accessible from outside of the container. In local deployment, to access the container ports from the host machine, one can publish a port from inside the container to the host machine.

Reviewer: Perhaps link out to the Docker network documentation (https://docs.docker.com/network/) for more information on how the Docker networking works?

```bash
docker run -p 8444:8444 origin
```

However, Pelican server still recognize `6ee28c5df997` as its hostname, causing a mismatch.

> In practice, there is usually a reverse-proxy service (such as Nginx, Traefic, etc.), to direct traffic from a human-readble domain name (e.g. https://example-origin.org) to your container (e.g. `6ee28c5df997:8444`). If Pelican server doesn't catch this information and uses it as its server address, there will be a mismatch when the server advertise itself to other Pelican services.

Reviewer 3: ```suggestion
In practice, there is usually a reverse-proxy service (such as Nginx, Traefic, etc.), to direct traffic from a human-readable domain name (e.g. https://example-origin.org) to your container (e.g. `6ee28c5df997:8444`). If Pelican server doesn't catch this information and uses it as its server address, there will be a mismatch when the server advertise itself to other Pelican services.
```

> Therefore, to keep things consistent, Pelican internally should use the same human-readble domain name as its web server's _external address_. To do so, configure `Server.ExternalWebURL` to the domain name the outside world uses to access the Pelican web service, e.g. https://example-origin.org:8444 for a production service or https://localhost:8444 for a local deployment.

Reviewer 3: ```suggestion
Therefore, to keep things consistent, Pelican internally should use the same human-readable domain name as its web server's _external address_. To do so, configure `Server.ExternalWebURL` to the domain name the outside world uses to access the Pelican web service, e.g. https://example-origin.org:8444 for a production service or https://localhost:8444 for a local deployment.
```

> Notice that there is a port number, `8444`, following the hostname in the URLs above. This is the default port for Pelican web server. When publishing ports from the container, you may choose a different port on the host machine to publish to. Two convential options are `443` or `8080`, which are default ports for a web service. If you publish `8444` port from the container to `443`, or `8080` on the host machine, i.e. `docker run -p 8444:443`, you may remove the port number from `Server.ExternalWebURL`, i.e. `https://example-origin.org`. For other ports, you need to leave the port number in the URL.

Reviewer 3: ```suggestion
Notice that there is a port number, `8444`, following the hostname in the URLs above. This is the default port for Pelican web servers. When publishing ports from the container, you may choose a different port on the host machine to publish to. Two conventional options are `443` or `8080`, which are default ports for a web service. If you publish `8444` port from the container to `443`, or `8080` on the host machine, i.e. `docker run -p 8444:443`, you may remove the port number from `Server.ExternalWebURL`, i.e. `https://example-origin.org`. For other ports, you need to leave the port number in the URL.
```

## Troubleshooting

### Error running Pelican Docker image on macOS with a M-series chip

When running Pelican Docker image on macOS with a M-series chip, you may get the following error message:

```console
docker: no matching manifest for linux/arm64/v8 in the manifest list entries.
See 'docker run --help'
```

> This is because Pelican Docker images are built for Linux/AMD64 architecture only. To fix the issue, run the image with `--platform linux/amd64` instead:

Reviewer 2: Let's add a note that this isn't a great idea for production servers because the x86 emulation makes things much less efficient.
